[블랙잭 1단계] 오이 미션 제출합니다.

cucumber99 opened the pull request:
# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

<!--
학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
-->

안녕하세요 둘리! 🫡
1차 리뷰 잘 부탁드립니다!!

- 딜러와 플레이어의 손패 (`List<Card>`) 를 감싸기 위해 `Hand`라는 일급 컬렉션을 구현했습니다. 또한 `Person`이라는 추상 클래스를 만들어 딜러와 플레이어 클래스를 구현하고 `Person` 내부에서 `Hand`를 관리하는데, 테스트 코드에서 특정 점수에 대한 테스트 등 여러 테스트에서 원하는 대로 카드 목록을 구성하기 위해서 플레이어와 딜러 객체를 생성할 때 `Hand` 객체를 생성자로 받을 수 있게끔 코드를 작성했습니다. 실제 도메인 로직에서는 `Deck`에서 카드를 무작위로 생성하여 전달해주기 때문에 위와 같이 코드를 작성했는데, 이 방식과 `Deck` 자체를 인터페이스로 추상화하여 테스트 코드에서 고정된 카드 목록을 전달해주는 방식 중 어느 방식이 더 바람직한지 둘리의 의견이 궁금합니다.🤔

- 복잡한 출력 형식을 맞추기 위해서 도메인 모델의 데이터를 UI 모델을 거쳐서 View에 전달되도록 하였습니다. 현재 UI 모델 내부에서 출력해야하는 문자열의 일부를 직접 구성해주고 있는데(열거형 클래스 타입에 따른 분기 처리 등), 이런 식으로 사용하는 것은 View의 역할까지 담당하게 되는 것인가요?

## 코드 리뷰 커뮤니케이션

<!--
리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
-->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

<!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

<!-- 필요하다면 스크린샷을 첨부하세요. -->

![]()

moondev03 committed:
chore: 코틀린 DSL 실습

moondev03 committed:
docs(README): 카드/덱 기능 구현 목록 작성

moondev03 committed:
feat: 숫자와 패턴을 가진 카드 생성 기능 구현

moondev03 committed:
feat: 고유한 52장의 카드를 가진 덱 생성 기능 구현

moondev03 committed:
feat: 요청한 수량에 맞게 카드를 뽑는 기능 구현

moondev03 committed:
feat: 패에 카드 추가 기능 구현

moondev03 committed:
feat: 플레이어 객체 구현

moondev03 committed:
feat: 패에 들고 있는 카드의 합을 구하는 기능 구현

moondev03 committed:
refactor: 플레이어/딜러로 분리

- Person 추상 클래스 사용해 중복 코드 제거

moondev03 committed:
refactor: Role 제거

moondev03 committed:
feat: 딜러의 점수와 각 플레이어의 점수를 비교하여 승패 여부를 판단하는 기능 구현

- 승패 여부는 Player가 관리

moondev03 committed:
chore: domain 패키징

- card
- person

moondev03 committed:
refactor: 결과 계산 로직 개선

- 플레이어 객체의 ResultState 제거

moondev03 committed:
feat: 플레이어 GameState 관리 기능 구현

moondev03 committed:
feat: Controller & View 구현

- 개행 & Depth 수정 필요

moondev03 committed:
feat: 패에 들고 있는 카드의 점수를 계산하는 기능 구현

moondev03 committed:
refactor: Controller에서 게임 상태 관련 기능 분리

moondev03 committed:
fix: Score 계산 로직 수정

moondev03 committed:
refactor: 게임 상태 갱신 & draw 로직 통합

moondev03 committed:
refactor: Controller Indent Depth 정리

moondev03 committed:
refactor: Player/Dealer State 분리

- GameRule Object 사용해 상수 관리
- Controller 도메인 로직 제거

moondev03 committed:
feat: 덱에 남은 카드가 없는 경우 검증 기능 구현

moondev03 committed:
chore: isCanDraw -> canDraw로 네이밍 변경

moondev03 committed:
refactor: PlayerState BLACK_JACE 상태 제거

moondev03 committed:
test: PlayerTest & DealerTest 추가

cucumber99 committed:
refactor: 점수 계산 기능 분리

cucumber99 committed:
refactor: draw 로직 개선

cucumber99 committed:
refactor: 점수 계산 로직 개선

cucumber99 committed:
refactor: Hand 클래스 개선

- copy 메소드 추가

cucumber99 committed:
refactor: 변수명 변경

- isHit -> hitFlag

cucumber99 committed:
refactor: 드로우 수량 로직 분리

cucumber99 committed:
refactor: 디렉터리 구조 변경

cucumber99 committed:
test: 딜러 상태 열거형 클래스 테스트

cucumber99 committed:
test: 플레이어 상태 열거형 클래스 테스트

cucumber99 committed:
refactor: 승패 여부 로직 조건 추가

- 딜러 버스트 케이스

cucumber99 committed:
feat: PersonUiModel 구현

cucumber99 committed:
feat: ResultUiModel 구현

cucumber99 committed:
refactor: OutputView 개선

- UI Model 반영
- 상수 분리

cucumber99 committed:
refactor: InputView 개선

- 상수 분리

cucumber99 committed:
refactor: draw 메소드 추상화

cucumber99 committed:
refactor: copy 로직 개선

cucumber99 committed:
style: 불필요한 개행 제거

cucumber99 committed:
refactor: 출력 로직 추가

- 딜러의 첫 카드만 공개합니다

cucumber99 committed:
refactor: 결과 관련 코드 개선

- GameResult가 결과를 상태로 갖습니다

cucumber99 committed:
refactor: CardPattern value 제거

moondev03 committed:
refactor: Hand 객체 불변성 유지하도록 변경

cucumber99 committed:
refactor: 불필요한 import 제거

hyemdooly commented on src/main/kotlin/blackjack/domain/card/Deck.kt:
```diff
@@ -0,0 +1,28 @@
+package blackjack.domain.card
+
+class Deck {
+    private val _cards = mutableListOf<Card>()
+    val cards: List<Card> get() = _cards.toList()
+
+    init {
+        _cards.addAll(generateDeck())
+        require(cards.size == DECK_SIZE) { INVALID_DECK_SIZE_ERROR_MESSAGE }
+    }
```
> 딜러와 플레이어의 손패 (List<Card>) 를 감싸기 위해 Hand라는 일급 컬렉션을 구현했습니다. 또한 Person이라는 추상 클래스를 만들어 딜러와 플레이어 클래스를 구현하고 Person 내부에서 Hand를 관리하는데, 테스트 코드에서 특정 점수에 대한 테스트 등 여러 테스트에서 원하는 대로 카드 목록을 구성하기 위해서 플레이어와 딜러 객체를 생성할 때 Hand 객체를 생성자로 받을 수 있게끔 코드를 작성했습니다. 실제 도메인 로직에서는 Deck에서 카드를 무작위로 생성하여 전달해주기 때문에 위와 같이 코드를 작성했는데, 이 방식과 Deck 자체를 인터페이스로 추상화하여 테스트 코드에서 고정된 카드 목록을 전달해주는 방식 중 어느 방식이 더 바람직한지 둘리의 의견이 궁금합니다.🤔

우선 저는 대체로 초기값을 전달해주는 편이에요. 이것만으로도 테스트가 쉬워진다고 생각하거든요. 카드를 전달해주거나, 카드 생성 방식을 전달해주거나요! 예시는 여러가지가 있을 것 같아요.
지금은 객체를 생성해서 어떤 함수를 실행한 다음에 assert가 돌아가도록 되어있지만, 초기값을 넣을 수 있다면 함수를 실행하지 않아도 되겠죠.
코드 전체적으로 초기값을 생성자에서 받기보다는, 빈 값을 넣고 액션을 하도록 되어있는데요. default parameter를 사용해볼 수도 있고, 빈 값으로 넣어주면 같은 동작을 하니, 각 도메인 모델들의 생성자를 한번 살펴보면 좋겠어요.
어떻게 건드리면 테스트코드가 쉬워질지 눈에 보일거에요!

hyemdooly commented on src/main/kotlin/blackjack/domain/card/Card.kt:
```diff
@@ -0,0 +1,15 @@
+    val number: CardNumber,
+    val pattern: CardPattern,
+) {
+    companion object {
+        fun create(
+            number: CardNumber,
+            pattern: CardPattern,
+        ): Card {
+            return Card(number, pattern)
+        }
+    }
```
팩토리 함수가 생성자랑 동일한 역할을 하는 것으로 보여요!

hyemdooly commented on src/main/kotlin/blackjack/uiModel/PersonUiModel.kt:
```diff
@@ -0,0 +1,42 @@
+import blackjack.domain.person.Person
+import blackjack.domain.person.Player
+
+data class PersonUiModel(
```
> 복잡한 출력 형식을 맞추기 위해서 도메인 모델의 데이터를 UI 모델을 거쳐서 View에 전달되도록 하였습니다. 현재 UI 모델 내부에서 출력해야하는 문자열의 일부를 직접 구성해주고 있는데(열거형 클래스 타입에 따른 분기 처리 등), 이런 식으로 사용하는 것은 View의 역할까지 담당하게 되는 것인가요?

사실 UiModel을 사용하는 경우는 실제 안드로이드 개발에서 자주 사용되긴 합니다. Data - Domain - Presentation 계층으로 나누게 되면서 각 계층의 모델들을 만들 수 있어요. **(지금 당장 몰라도 되는 내용이니 그렇구나 하고 넘어가주세요.ㅎㅎ)**
우선 저는 Ui Model이 도메인이 아니라고 생각해요. 지금 구조에서는 View의 역할에 가깝다고 생각합니다.
이 모델을 그대로 안드로이드로 옮겨야한다면, 안드로이드에서는 실제 이미지를 출력하기도 할거고, 색깔이 입혀지기도 할거고, 더 다양한 UI 요소들이 포함되겠죠?
하지만 그렇다고 해서 잘못된 구조라고 생각하지 않습니다. 지금 요구사항에서는 View가 Domain 모델을 알아도 되지만, 분리를 한다고 해서 잘못됐다고 생각하지는 않아요.
대신 앞으로 미션 구현하면서 이러한 Ui Model들이 View에서 사용될 수 있도록, View가 Domain 모델을 사용하지 않도록 일관적인 코드를 작성하는 것이 중요하겠어요!

hyemdooly commented on src/main/kotlin/blackjack/domain/ScoreCalculator.kt:
```diff
@@ -0,0 +1,33 @@
+        return adjustAceValues(sum, values)
+    }
+
+    private fun getCardValue(card: Card): Int {
+        if (card.number == CardNumber.ACE) return GameRule.ACE_OTHER_SCORE
+        return card.number.value
+    }
+
+    private fun adjustAceValues(
+        sum: Int,
+        values: List<Int>,
+    ): Int {
+        var total = sum
+        val aceCount = values.count { it == GameRule.ACE_OTHER_SCORE }
+
+        repeat(aceCount) {
+            if (total > GameRule.BLACKJACK_SCORE) {
+                total -= GameRule.ACE_BASE_SCORE
+            }
+        }
+        return total
+    }
```
이런 판단 로직을 `CardNumber`에 옮길 수 있지 않을까요? (특히, `ACE`가 1이 되어야하는지, 11이 되어야하는지 판단 로직)
`enum class`에서도 function 선언이 가능하니까요!

hyemdooly commented on src/main/kotlin/blackjack/domain/person/Dealer.kt:
```diff
@@ -0,0 +1,20 @@
+import blackjack.domain.card.Deck
+import blackjack.domain.state.DealerState
+
+class Dealer(hand: Hand) : Person(hand) {
+    init {
+        gameState = DealerState.FIRST_TURN
+    }
+
+    constructor() : this(hand = Hand())
```
여기서 부생성자가 필요할까요?

cucumber99 replied:
이 부분이 위에서 언급했던 점입니다.
특정 카드들을 플레이어나 딜러가 가지고 있을 때의 상황을 테스트하기 위해서 `Hand`를 생성자로 전달하거나, 전달하지 않을 경우 내부에서 기본으로 생성하게 됩니다.
때문에 외부에서 부 생성자로 `Hand`를 받아올 경우 외부와의 참조를 끊기 위해 `Hand`에 `copy`를 선언하고 그대로 사용했습니다.

hyemdooly replied:
아하 Default value를 사용할 수 있을 것 같아서 드린 리뷰였어요! [공식문서](https://kotlinlang.org/docs/classes.html#constructors)

hyemdooly commented on src/main/kotlin/blackjack/domain/state/DealerState.kt:
```diff
@@ -0,0 +1,20 @@
+    FINISH(true),
+    ;
+
+    companion object {
+        fun from(dealer: Dealer): DealerState =
+            when {
+                dealer.cards().size < GameRule.FIRST_TURN_DRAW_AMOUNT -> FIRST_TURN
+                dealer.score() > GameRule.DEALER_ADDITIONAL_DRAW_BASE_SCORE -> FINISH
+                else -> HIT
+            }
+    }
```
`State`들이 객체를 받기 보다는 실제 값들(카드 개수, 스코어)을 받아서 판단하면 어떨까요?
여기서 관심사는 사람보다는 사람이 가지고 있는 것들이라고 생각해요!

hyemdooly commented on src/main/kotlin/blackjack/domain/person/Person.kt:
```diff
@@ -0,0 +1,30 @@
+
+    fun score(): Int = calculator.calculate(cards())
+
+    protected fun getDrawAmount(state: PersonState): Int {
+        if (gameState == state) {
+            return GameRule.FIRST_TURN_DRAW_AMOUNT
+        }
+        return GameRule.HIT_DRAW_AMOUNT
+    }
```
state가 같으면 2장, 같지 않으면 1장 이런 로직을 갖고 있는데요, `State`에 따른 뽑을 수 있는 카드 장수를 선언하면 어떨까요?

cucumber99 replied:
상태에 따른 분기 처리, 예외 처리까지 할 수 있을 것 같아요.
좋은 의견 감사합니다😊

hyemdooly commented on src/main/kotlin/blackjack/uiModel/PersonUiModel.kt:
```diff
@@ -0,0 +1,42 @@
+            )
+        }
+
+        private fun Card.toUiString(): String = "${number.toUiString()}${pattern.toUiString()}"
```
확장 함수 사용 👍

hyemdooly commented on src/main/kotlin/blackjack/uiModel/PersonUiModel.kt:
```diff
@@ -0,0 +1,42 @@
+
+        private fun CardNumber.toUiString(): String =
+            when (this) {
+                CardNumber.ACE -> "A"
```
상수화를 해주면 어떨까요? (이하 동일)

cucumber99 replied:
페어와 이 부분에서 해당 상수들을 `GameRule`에서 관리하기 애매하다고 생각해서 바꾸지 않은 부분인데, 개선해보겠습니다!

hyemdooly commented on src/test/kotlin/blackjack/domain/card/DeckTest.kt:
```diff
@@ -0,0 +1,30 @@
+    fun `고유한 52장의 카드를 가지고 있어야 한다`() {
+        val deck = Deck()
+
+        deck.cards.size shouldBe 52
```
테스트 코드에 kotest의 DSL을 활용해주셨군요 👍

hyemdooly commented on src/test/kotlin/blackjack/domain/card/CardTest.kt:
```diff
@@ -0,0 +1,14 @@
+    fun `숫자와 패턴을 가진 카드를 생성한다`() {
+        val card = Card.create(CardNumber.ACE, CardPattern.HEART)
+
+        assertThat(card.number).isEqualTo(CardNumber.ACE)
+        assertThat(card.pattern).isEqualTo(CardPattern.HEART)
```
`assert`를 한 함수에 여러번 쓰면 어떤 불편함이 생길 수 있을까요?

cucumber99 replied:
`assertThat`중 하나라도 검증에 실패하면 이후 `assertThat`이 실행되지 않는 문제가 있었네요.
`kotest`나, `assertAll`을 활용해보겠습니다!

hyemdooly reviewed (changes requested):
오이 안녕하세요! 이번 미션을 함께 하게 된 둘리입니다. 반가워요! 😄 
전체적으로 객체를 잘 나눠주셨고, 상속을 사용해주신 부분, 상태를 적용하려고 하신 부분이 좋았습니다. 💯 
질문에 대한 답과 함께 추가로 고민해볼만한 내용들로 코멘트를 달아드렸어요.
더 궁금한 점이 있으시다면 코멘트 또는 DM 주세요! 👍

cucumber99 committed:
refactor: 카드 테스트 코드 개선

- assertSoftly를 이용해 여러 개의 테스트를 동시에 수행합니다

cucumber99 committed:
refactor: 게임 상태 통합

- 딜러와 플레이어가 같은 타입의 상태를 사용합니다

cucumber99 committed:
refactor: Participant 객체 리팩터링

- Person -> Participant로 이름을 변경합니다.
- Participant 내부에서 카드 목록을 직접 관리합니다.
- 상태에 따른 분기 처리를 세분화합니다.

cucumber99 committed:
refactor: Deck 객체 리팩터링

- 생성자로 카드 목록을 받습니다.
- 카드 개수에 따른 유효성 검증을 하지 않습니다.

cucumber99 committed:
refactor: Card 객체 리팩터링

- data class로 변경합니다.
- 팩토리 메소드 대신 생성자를 사용합니다.

cucumber99 committed:
refactor: Player 객체 리팩터링

cucumber99 committed:
refactor: Dealer 객체 리팩터링

cucumber99 committed:
refactor: 게임 결과 로직 개선

cucumber99 committed:
refactor: GameResult 객체 리팩터링

- 테스트 코드를 변경합니다.
- 승패 판단 로직을 개선합니다.

cucumber99 committed:
refactor: 문자열 상수화

cucumber99 committed:
refactor: 점수 계산 테스트 코드 개선

cucumber99 committed:
chore: 패키지명 변경

cucumber99 committed:
test: 테스트 코드 수정

cucumber99 committed:
feat: 상수 추가

cucumber99 committed:
feat: Game 객체 구현

- 게임의 전반적인 흐름을 제어합니다.

cucumber99 committed:
refactor: 입력 로직 개선

cucumber99 committed:
refactor: 출력 로직 개선

cucumber99 committed:
refactor: Controller 로직 변경

cucumber99 committed:
refactor: 파일명 변경

cucumber99 committed:
test: Game 객체 테스트 코드

cucumber99 committed:
feat: 게임 결과 로직 변경

cucumber99 commented:
안녕하세요 둘리!🫡
이전 리뷰 감사했습니다!
리팩터링을 진행하며 최대한 피드백을 반영하되, 구조적으로 개선해보았습니다.

- 게임의 흐름을 담당하는 `Game` 객체 구현
- `Player`와 `Dealer` 사이의 공통 로직 제거
- `Hand` 객체 제거
- `GameState` 개선
- 로직 개선 등..

컨트롤러에서 도메인 로직을 최대한 없애기 위해 `Game` 객체를 구현하였는데, 
오히려 테스트의 복잡성이 증가하는 등 여러 문제가 생긴 것 같습니다...
이번 리뷰도 잘 부탁드립니다🥲

hyemdooly commented on src/main/kotlin/blackjack/domain/participants/Player.kt:
```diff
@@ -0,0 +1,12 @@
+
+class Player(
+    val name: String,
+    private val checkHit: (String) -> Boolean,
```
출력 때문에 String에 이름을 넣기 위해 이렇게 된 람다식으로 보입니다.
뷰를 제외하고 봤을 때, 이런 구조는 자연스러운 구조일까요? 🤔

cucumber99 replied:
사용자의 입력에 따라 플레이어의 HIT 여부가 결정되기 때문에 로직을 더욱 간결화하기 위해서 위와 같은 구조를 구현했습니다.
근데 둘리가 말한 몇 가지 버그 중 하나가 해당 구조때문에 발생한다는걸 알아버렸어요..🥲
또한 플레이어 객체가 `View`의 메소드에 의존하고 있기 때문에 문제가 될 수 있을 것 같아요.
해당 부분은 개선하겠습니다!

hyemdooly commented on src/main/kotlin/blackjack/controller/BlackJackController.kt:
```diff
@@ -0,0 +1,69 @@
+        return Game(dealer, players)
+    }
+
+    private fun setGame(game: Game) {
```
이름은 `setName`이지만 하는 일은 출력이군요 🤔

cucumber99 replied:
네이밍을 명확한 의미를 전달할 수 있게끔 바꿔보겠습니다!

hyemdooly commented on src/main/kotlin/blackjack/domain/card/Deck.kt:
```diff
@@ -0,0 +1,25 @@
+        return _cards.removeFirst()
+    }
+
+    private fun generateDeck(): List<Card> = CardPattern.entries.flatMap(::createCard).shuffled()
+
+    private fun createCard(cardPattern: CardPattern): List<Card> {
+        return CardNumber.entries.map { cardNumber -> Card(cardNumber, cardPattern) }
+    }
```
`Deck`에서 카드를 생성하는 것은 자연스러운 액션일까요?

cucumber99 replied:
저는 `Deck`이 게임에서 사용될 카드들을 관리하므로, 생성에 대한 책임 또한 가지고 있어도 된다고 생각했습니다!
만약 여러 가지의 덱을 사용하거나, 덱 생성 규칙이 달라진다면 덱 자체를 추상화하거나, 생성 방식을 추상화하여 사용할 수 있을 것 같습니다.
둘리의 생각은 어떠신가요?🤔

hyemdooly replied:
저는 카드들을 외부에서 주입받는 형태를 사용할 것 같아요. 🤔 
덱은 누군가가 카드를 뽑을 수 있도록 관리하는 역할은 맞지만, 생성을 한다고 생각하지는 않아요.
카드가 비었다고 해서 새로운 덱을 생성해주는 것도 어색해요.
DeckTest의 코드를 보시면, 빈 덱을 만들기 위해서 카드 한장을 넣어주고 두번 뽑도록 되어있더라구요. (추가로 리뷰 드릴게요!)
처음부터 빈 값을 세팅할 수 있다면 더 자연스럽지 않을까요?

hyemdooly commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -0,0 +1,67 @@
+        println(DRAW_STATUS_MESSAGE.format(participantsUiModel.name, participantsUiModel.cards.joinToString(DELIMITER)))
+    }
+
+    fun printDealerDrawMessage(flag: Boolean) {
```
flag말고 더 명시적인 이름이 있을까요?

cucumber99 replied:
딜러가 카드를 받았다는 것을 전달할 수 있게끔 이름을 변경해보겠습니다!

hyemdooly commented on src/main/kotlin/blackjack/domain/Game.kt:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    private fun handOutCard(participant: Participant) {
+        repeat(participant.getDrawAmount()) {
+            val card = dealer.draw()
+            participant.addCard(card)
+        }
```
이게 무슨 로직일까 생각했는데 딜러가 뽑아서 플레이어에게 주고있는 것이군요
결국 플레이어가 카드를 뽑고 있는 것이니, 각자 덱에서 카드를 뽑는다 생각하면 더 단순해지지 않을까요?

cucumber99 replied:
특정 덱에서 카드를 뽑는 책임 자체를 딜러에게 위임하기 위해서 위와 같이 코드를 작성했어요.
다만 여기서 궁금한 점이, `dealer.draw()`는 결국 `deck.draw()`를 반환할 뿐, 추가적인 동작은 하지 않는데요.
이렇게 되면 카드를 드로우하는 책임은 최종적으로 `Deck`에게 있는건가요?

hyemdooly replied:
카드를 뽑아서 리턴하고 있는건 deck이고 뽑은 카드를 리턴하고 있는 것도 dealer인데,
deck에서 뽑은 카드를 그대로 dealer가 던져준다고 해서 dealer에게는 책임이 없다고 할 수 있을까요?
크루들과 함께 대화해보면 좋을 것 같네요!

hyemdooly commented on src/main/kotlin/blackjack/domain/state/GameState.kt:
```diff
@@ -0,0 +1,25 @@
+    BLACKJACK,
+    ;
+
+    companion object {
+        fun from(participant: Participant): GameState {
+            return when {
+                participant.score() > GameRule.BLACKJACK_SCORE -> BUST
+                participant.score() == GameRule.BLACKJACK_SCORE && participant.hand.size == GameRule.FIRST_TURN_DRAW_AMOUNT -> BLACKJACK
+                participant is Dealer && participant.score() > GameRule.DEALER_ADDITIONAL_DRAW_BASE_SCORE -> STAY
+                else -> HIT
+            }
+        }
+    }
```
상태를 적용할 것이라면 상태 계산 로직은 팩토리 함수보다는, 객체 각자가 들고있는 것이 좋지 않을까요?
객체가 자기 스스로 상태를 판단할 수 있어 보입니다!
만약 다른 포지션이 추가되거나, 추가 룰이 적용된다면 `from` 함수는 더 거대해지지 않을까요?

cucumber99 replied:
객체가 스스로 상태를 판단할 수 있게끔 하는 것은 저도 매우 바람직하다고 생각합니다😊
다만, 해당 로직이 `Participants` 내부에 있다고 가정했을 때 포지션이 추가되거나 추가 룰이 적용되어 그에 따라 로직이 커지는 것은 달라지지 않을 것 같다고 생각해요.
우선 객체 스스로 상태를 판단할 수 있도록 코드를 개선해보겠습니다!

hyemdooly replied:
'추상화'를 고민해보시면 좋을 것 같아요!

hyemdooly commented on src/main/kotlin/blackjack/domain/participants/Dealer.kt:
```diff
@@ -0,0 +1,13 @@
+import blackjack.domain.state.GameState
+
+class Dealer(
+    private val deck: Deck,
+) : Participant() {
+    fun draw(): Card = deck.draw()
```
딜러가 뽑아서 나눠주는 구조라 deck을 가지고 있는 것이었군요!
개인적인 생각으로는, 뽑을 때 deck을 파라미터로 받는 것은 어떨까 싶습니다.
그렇다면 명시적으로 어떤 덱에서 카드를 뽑는지 표현되지 않을까! 라는 생각이에요. 오이의 생각도 궁금합니다!

cucumber99 replied:
현실에서의 블랙잭 게임을 생각해보면, 딜러가 정확히 하나의 덱을 가지고 있으면서 각 플레이어와 자신에게 카드를 분배하는 것이 적절하다고 생각했습니다.
때문에 `Deck` 객체를 생성자로 받고, `draw()`를 호출하게끔 코드를 작성하였습니다.
`draw()` 메소드가 `Deck` 객체를 인자로 받게 된다면, 한 게임에서 하나의 덱이 아니라 다른 종류의 덱을 사용할 수도 있을 것 같다고 생각했습니다.

hyemdooly reviewed (changes requested):
오이 안녕하세요! 많은 고민을 하신 것이 보이네요. 💪 

- `Game`이라는 친구가 생겼는데 이 친구가 다 하려고 해서 더 읽기 어려워진 느낌을 받았어요. 😭 
  - `Controller`, `Game`에서 타고 들어가는 함수가 너무 많다고 느꼈어요. 공통된 로직이 반복되어 분리한 함수가 아닌 이상, 함수를 너무 쪼개개되면 흐름을 따라가기 더 어려울 수 있습니다. (onHit 함수... 오래 쳐다봤어요...👀 )
- 한 객체에 데이터를 넣어놓고 꺼내쓰는 경우가 보이는데, 객체가 스스로 하게끔 할 수 있을지 고민해보면 좋을 것 같아요.
- 객체 안에서 스스로 생성하는 것들이 종종 있는데, 이 점이 테스트를 더 번거롭게 만들지는 않은지 고민해보면 좋을 것 같아요!

그리고 실행시켜보니 몇 가지 버그들이 나오는데, **블랙잭 게임 흐름을 정리해보고 단순하게 생각해보시는 것이 좋을 것 같아요. 어려우면 상태 패턴도 제거하셔도 좋습니다.**
처음 딜러와 플레이어가 2장씩 뽑는다 부터, 플레이어들의 버스트 체크 + 추가 카드 여부, 딜러의 추가 카드 여부, 그리고 마지막 점수 계산... 게임 순서는 정해져있고 생각보다 단순한 게임이니 차근차근 정리해보세요!

다음 요청에는 되도록 merge하도록 할게요! 화이팅입니다! 💪

hyemdooly commented:
아침에 생각해보니.. UiModel과 상태 패턴을 제거하고, 뷰에서 도메인 모델을 알아도 되니 한번 덜어내보시는 것도 추천드릴게요!💪
오히려 그런 패턴이 생각을 많게 해서 더 복잡하게 만들고 있는 것 같아서요🤔
일단 쉽게쉽게 필요한 것에만 집중해서 작성해보는건 어떠신가요?!

cucumber99 committed:
refactor: 메소드 이름 변경

cucumber99 committed:
refactor: 상태 클래스 제거

cucumber99 committed:
refactor: Participant 클래스 리팩터링

cucumber99 committed:
refactor: ScoreCalculator 클래스 리팩터링

cucumber99 committed:
refactor: 테스트 코드 수정

cucumber99 committed:
refactor: Player 클래스 리팩터링

cucumber99 committed:
refactor: Dealer 클래스 리팩터링

cucumber99 committed:
refactor: 기타 코드 리팩터링

- Player, Dealer 변경에 따라 리팩터링합니다.

cucumber99 committed:
refactor: Game 클래스 리팩터링

cucumber99 committed:
refactor: Deck 클래스 리팩터링

cucumber99 committed:
refactor: Controller 리팩터링

cucumber99 commented:
안녕하세요 둘리!🫡
둘리의 리뷰를 받고 상당 부분 개선할 수 있었어요.

기존의 코드는 `GameState`로 플레이어와 딜러의 상태를 나타내면서, 카드가 추가될 때 마다 점수 계산 메소드를 호출하고 상태를 갱신해주어야 했습니다.
프로그램의 규모가 작기 때문에 큰 문제가 되지는 않았지만, 비효율적이라고 생각했어요.
또한, `GameState`에 따라 동작을 분기 처리해야 했는데, 상태를 객체가 갖고 있으면서 상태 비교는 외부에서 하는 것이 어색했어요.
최대한 로직의 복잡한 부분은 걷어내고 객체가 스스로 행동하게끔 했습니다.

> Deck에서 카드를 생성하는 것은 자연스러운 액션일까요?

다만, 이 피드백은 좀 더 얘기를 나눠보고 싶어요.
이 부분에 대해 둘리의 의견을 더욱 자세하게 들려주시면 감사하겠습니다!

이번 리뷰도 잘 부탁드립니다.😊

hyemdooly commented on src/test/kotlin/blackjack/domain/card/DeckTest.kt:
```diff
@@ -0,0 +1,33 @@
+import org.junit.jupiter.api.Test
+
+class DeckTest {
+    @Test
+    fun `카드를 한 장 뽑는다`() {
+        // Given
+        val cards = MutableList(3) { Card(CardNumber.ACE, CardPattern.HEART) }
+        val deck = Deck(cards)
+
+        // When
+        deck.draw()
+
+        // Then
+        deck.cards.size shouldBe 2
+    }
```
given/when/then에 어떤 행위를 하는지 다시 한번 살펴보면 좋겠어요.
제가 보기에 여기서 when은, given에 가까워보여요. 빈 덱을 만들기 위함이니까요!
이것도 사실 빈 값으로 세팅했으면 바로 draw()하면 됐을텐데 라는 생각이 드네요.

hyemdooly commented on src/main/kotlin/blackjack/controller/BlackjackController.kt:
```diff
@@ -0,0 +1,65 @@
+        outputView.printDrawMessage(toParticipantsUiModel(game.dealer, game.players))
+    }
+
+    private fun askHit(game: Game) {
+        game.askHit(
+            decideHit = { inputView.getFlag(it) },
+            onHit = { outputView.printDrawStatus(ParticipantsUiModel.create(it)) },
+        )
+    }
```
어디까지가 게임의 역할이고, 컨트롤러의 역할일까요?

사실 저는 지금 구조에서 Game 존재의 이유를 잘 모르겠어요. 진행 중에 input함수와 output함수가 필요해서 람다를 넘기고 있는데, 출력하기 위해 람다함수의 파라미터가 객체가 들어가고 있기도 하고요. (출력이 아니었다면 들어가지 않았을 파라미터인 것 같아요.)
오히려 분리해서 더 가독성이 떨어지는 것 같아요. 🤔 
코드를 읽다보면 이 람다가 왜 필요했지?하고 다시 컨트롤러를 보게 되네요...

cucumber99 replied:
제가 `Game` 객체를 구현하게 된 가장 큰 이유는 컨트롤러에서 도메인 로직을 분리하기 위함이었어요.

- 게임 시작 시 플레이어와 딜러에게 카드 2장 드로우
- 이후 플레이어와 딜러의 드로우 여부 확인
 
위의 두 로직들이 컨트롤러에 노출되어 있는 것이 어색하다고 느껴졌습니다.
다만, 플레이어가 드로우 가능 상태일 때 사용자의 입력에 따라 동작이 달라지기 때문에 메소드를 인자로 전달하는 방식으로 코드를 작성했어요.

이 부분에서 둘리의 생각이 궁금합니다.🫡
`Game` 객체를 사용하지 않고, 컨트롤러에 도메인 로직이 노출되어도 괜찮을까요?
혹은 컨트롤러에 도메인 로직이 노출될 수 밖에 없는 상황이라면, 설계가 잘못된 것인가요?

hyemdooly replied:
어디까지 도메인 로직으로 봐야할까요? 플레이어의 한 턴이 끝나고나서 출력하는 것도 컨트롤러가 아닌 게임에서 실행해야할까요?
이 리뷰가 Game을 사용하지 말라는 이야기는 아니었습니다. 🤔

게임 시작 시 플레이어와 딜러에게 카드 2장씩 드로우하는 것은 게임의 규칙이니 게임에 있을 수 있다고 생각해요.
하지만 이후에 카드를 더 뽑을 것이냐 말것이냐와 같은, 뷰에서 받아오는 것까지 Game에 숨겨버린다면 복잡성이 더 커진다고 생각합니다.
`decideHit`, `onHit` 함수 둘 다 입력 받아오거나 출력하는 함수인데 이 함수들까지 Game에서 실행하려니 복잡해지는 것 같아요.

cucumber99 replied:
맞습니다. 도메인 로직이 뷰에 의존하는 형태가 되어버려서 좋은 형태는 아니라고 생각해요.
처음에는 `어차피 사용자의 응답에 따라 동작이 달라지는데, 상관 없지 않은가?` 라고 생각해서 함수를 파라미터로 전달했는데, 둘리의 리뷰를 보고 나서 다른 크루들에게도 이야기를 해보니 좋은 점보다 좋지 않은 점이 더 많다고 느껴졌어요.

지금 당장은 Game의 문제를 해결할 방법이 떠오르지 않아서, 우선은 Game 객체를 사용하지 않게끔 하겠습니다.
좋은 의견 감사합니다!😊

hyemdooly replied:
오 제 의견이 잘못 전달이 된 것 같은데요! 제가 모호하게 말씀을 드린 것 같네요 🙏 
**람다 함수로 전달하는 것은 도메인이 뷰에 의존하지 않는 것이 맞습니다. 뷰의 함수를 참조하는 것이 아니라서요!**

저도 이 이후로 조금 더 생각을 해봤는데, 오이 의견대로 우선 Game을 사용하되 함수를 너무 타고 들어가지 않게끔 단순하게 만들면 좋아지지 않을까 싶어요!!

hyemdooly commented on src/main/kotlin/blackjack/const/GameRule.kt:
```diff
@@ -0,0 +1,22 @@
+package blackjack.const
+
+object GameRule {
```
이 상수들을 관련된 도메인 모델로 옮겨볼 수 있지 않을까요?

cucumber99 replied:
하나의 상수가 여러 곳에서 사용되는 경우가 있어서 우선 `GameRule`에 모아놨습니다.
적절한 곳에 위치할 수 있도록 옮겨보겠습니다!

hyemdooly commented on src/main/kotlin/blackjack/view/InputView.kt:
```diff
@@ -0,0 +1,35 @@
+        return readUserInput().split(DELIMITER).map { it.trim() }
+    }
+
+    fun getFlag(name: String): Boolean {
```
함수 이름도 바뀌면 좋겠어요!

hyemdooly commented on src/main/kotlin/blackjack/domain/participants/Participant.kt:
```diff
@@ -0,0 +1,38 @@
+
+    fun score(): Int = ScoreCalculator.calculate(hand)
+
+    fun getDrawAmount(): Int {
+        if (hand.isEmpty()) {
+            return GameRule.FIRST_TURN_DRAW_AMOUNT
+        }
+
+        if (canHit()) {
+            return GameRule.HIT_DRAW_AMOUNT
+        }
+
+        throw IllegalArgumentException("[ERROR] 카드를 뽑을 수 없습니다.")
+    }
```
더 뽑을 수 없다면 0을 리턴하는 것이 자연스럽지 않을까요?

cucumber99 replied:
예외를 발생시키는 것보다 둘리가 말한 방식이 더 자연스러울 것 같아요😊

hyemdooly commented on src/test/kotlin/blackjack/domain/GameResultTest.kt:
```diff
@@ -0,0 +1,152 @@
+        val result = GameResult.create(blackjackDealer, listOf(player))
+
+        // Then
+        result.winStatus[player] shouldBe ResultState.DRAW
```
테스트함수 이름과 결과가 일치하지 않는 것 같아요. 🤔

cucumber99 replied:
미처 확인하지 못한 것 같습니다😓

hyemdooly commented on src/test/kotlin/blackjack/domain/GameTest.kt:
```diff
@@ -0,0 +1,143 @@
+    @Test
+    fun `게임 시작 시 딜러와 플레이어에게 카드를 분배한다`() {
+        // Given
+        val player = Player("test")
+        val game = Game(dealer, listOf(player))
```
사실 저에게는 생성자에 다른 객체들을 넘기고, init 블럭에서 넘어온 객체들을 변화시키는 구조가 어색하게 느껴져요. Game을 생성하고 테스트 코드 중에서는 아예 사용하지도 않고 있어서요.

테스트 코드만 봐도 어떤 역할을 하는지 명시적이지 않고, 무슨 역할인지 알기가 어려운 것 같아요. 😅

cucumber99 replied:
게임이 시작했을 때 딜러와 플레이어들에게 카드를 2장씩 나눠주는 기능을 메소드로 분리했습니다!

hyemdooly commented on src/test/kotlin/blackjack/domain/GameResultTest.kt:
```diff
@@ -0,0 +1,152 @@
+                Card(CardNumber.JACK, CardPattern.HEART),
+                Card(CardNumber.EIGHT, CardPattern.HEART),
+            )
+        dealerCards.forEach { card ->
+            dealer.addCard(card)
+        }
```
draw와 마찬가지로 여기도 생성자로 주입할 수 있으면 이렇게 일일히 카드를 추가하지 않아도 될 것 같은데 말이죠 🤔

cucumber99 replied:
딜러나 플레이어가 생성자로 `List<Card>`를 받는다는 말씀이신가요?

cucumber99 replied:
추가로 궁금한 점이 있어요!
딜러와 플레이어 내부의 카드 목록은 `Deck`에서 드로우된 카드를 추가할 때 변경되며, 실제 게임에서는 생성자로 직접 전달할 일이 없다고 생각해요. (불변 객체를 사용하는 경우는 생성자로 전달해야 할 것 같아요)

테스트 코드에서 카드를 직접 추가하지 않고 생성자로 주입하기 위해서는 도메인의 코드가 바뀌어야 하는데요.
테스트를 위해 도메인을 변경하는 것은 바람직한가요?
아니면 둘리가 생각하는 다른 의도가 있으신가요?👀

hyemdooly replied:
테스트가 용이하도록 설계를 개선하는 것은 도메인을 변경하는 것일까요?

hyemdooly reviewed (approved):
안녕하세요, 오이! 1단계 미션 고생 많으셨습니다!!
2단계 미션은 리뷰 드린 것들과 함께, 컨트롤러-게임의 역할을 더 고민해보시면 좋겠어요.
더 덜어낼 필요는 없는지, 꼭 필요한건지 생각해보시면 좋겠습니다.
궁금한 점은 코멘트를 더 달아주시거나, 2단계 PR에 함께 기재해주세요. 2단계 미션도 화이팅입니다! 🏃‍♂️
